wasmtime-wasi/src/syscalls.rs

```diff
@@ -1,45 +1,42 @@
-use crate::host::{argv_environ_values, fd_prestats, fd_table};
-use crate::instantiate::WASIState;
+use crate::translate;
+
 use cranelift_codegen::ir::types::{Type, I32, I64};
-use host;
-use host_impls;
-use std::{mem, ptr, slice, str};
-use translate::*;
-use wasm32;
+use wasi_common::ctx::{VmContext, WasiCtx};
+use wasi_common::host;
+use wasi_common::hostcalls;
+use wasi_common::wasm32;
 use wasmtime_runtime::VMContext;
 
-fn str_for_trace<'str>(ptr: *const i8, len: usize) -> Result<&'str str, str::Utf8Error> {
-    str::from_utf8(unsafe { slice::from_raw_parts(ptr as *const u8, len) })
-}
-
-fn return_encoded_errno(e: host::__wasi_errno_t) -> wasm32::__wasi_errno_t {
-    let errno = encode_errno(e);
-    trace!("    -> errno={}", wasm32::strerror(errno));
-    errno
+struct VmCtx {
+    vmctx: *mut VMContext,
 }
 
-unsafe fn get_curfds(vmctx: *mut VMContext) -> *mut fd_table {
-    (&mut *(&mut *vmctx)
-        .host_state()
-        .downcast_mut::<WASIState>()
-        .unwrap()
-        .curfds) as *mut fd_table
-}
+impl VmContext for VmCtx {
+    fn as_wasi_ctx(&self) -> *const WasiCtx {
```

Comment:
Here, the story behind `*const` and `*mut` is convoluted. In `lucet`, `WasiCtx` is stored inside a `HashMap` and originally, in `lucet-wasi`, it was accessed in `VmCtx` via `get_embed_ctx` which returns `Ref<T>` (or `RefMut<T>` for the mutable version). This in conjunction with the fact that `dec_ptr` in `lucet-wasi` requires only `&self` (unlike in `wasmtime-wasi` where we require `&mut self`), made it possible to operate with `ref`s in the hostcalls impls. After I've modified the signature of `dec_ptr` to take `&mut self`, I was unsurprisingly greeted with loads of complaints from our friend the borrow-checker (and rightly so) that I cannot borrow mutably and immutably at the same time. This needs to be fixed here, and in `wasi-common` first and foremost I believe, as it's the main blocker to having the hostcalls impls implemented using majority of safe Rust.

Reply:
This one is also outdated since we've decide to go a different route with `wasi-common` whereby the hostcalls provided by the lib expect either a reference to `WasiCtx`, or a mutable reference to guest's memory, or both.